### PR TITLE
Chore/report auto generator amendments on quarterly reports

### PR DESCRIPTION
**Pull request by Sepehr-Sobhani**
[ZH Card](https://app.zenhub.com/workspaces/climate-action-secretariat-60ca4121764d710011481ca2/issues/bcgov/cas-cif/1154)

### CODE REVIEW

**Review by BCerki — DISMISSED** [review]
Thanks for the demo, looks great!
This question is beyond the scope of the card, but--when I was manually testing I accidentally put the end date before the start date and tried to generate. The only front-end feedback you get is the error alert at the top of the screen, which doesn't explain where you've made the mistake (either on the budgets or TEIMP form, not on the quarterly reports form where you are). @nanyangpro, is this a problem? Thanks!

File: app/components/ReportingRequirement/ReportGenerator.tsx [hunk 1]
```diff
@@ -16,6 +17,12 @@ interface Props {
   startDateObject: DateFieldObject;
   endDateObject: DateFieldObject;
   readonly: boolean;
+  mutationFunction: (args: {
+    variables:
+      | generateQuarterlyReportsMutation$variables
+      | generateReportsMutation$variables;
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/components/ReportingRequirement/ReportGenerator.tsx:23]: Haha, I see what you mean. It's a prettier thing and not a typescript thing. I've put `var1 | var2` but prettier turned it into
```
| var1
| var2
```
I found a related issue here: https://github.com/prettier/prettier/issues/4168

File: schema/deploy/mutations/generate_quarterly_reports.sql [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  start_date_year int := extract(year from $2);
+  report_interval_start_date timestamptz;
+begin
+  if $2 >= $3 then
+      raise exception 'start_date must be before end_date';
+  end if;
+
+  -- choosing the closest quarter start date from a list of possible dates as a temporary table
+  select * from generate_series(make_date(start_date_year, 01, 05), make_date(start_date_year + 1, 01, 05), '3 month'::interval) as quarterly_report_due_date
+    where $2 <= quarterly_report_due_date limit 1 into report_interval_start_date;
```
Comment (tmastrom) [inline schema/deploy/mutations/generate_quarterly_reports.sql:18]: Very cool, didn't know you could select the first instance of match from a list like this

**Review by tmastrom — APPROVED** [review]

**Review by pbastia — APPROVED** [review]
very fine work

**Pull request merged**